File: src/python/day11.py
from functools import cache

example_input = """
svr: aaa bbb
aaa: fft
fft: ccc
bbb: tty
tty: ccc
ccc: ddd eee
ddd: hub
hub: fff
eee: dac
dac: fff
fff: ggg hhh
ggg: out
hhh: out
"""

def parse_input(input) -> dict[str, list[str]]:
  g = {}
  for line in input.strip().split("\n"):
    v, ulist = line.split(": ")
    g[v] = ulist.split(" ")
  return g

def bfs(g: dict[str, list[str]]) -> int:
  q = []
  q.append(["you", 0])
  paths = []
  while q:
    v, d = q.pop(0)
    if v in q:
      continue
    if v == "out":
      paths.append(d)
      continue
    for u in g[v]:
      q.append([u, d + 1])
  return len(paths)

def dfs(g: dict[str, list[str]]):
  @cache
  def dfs_util(v: str, dac: bool, fft: bool, end: str) -> int:
    if v == end:
      if dac and fft:
        return 1
      return 0
    if v == "dac":
      dac = True
    if v == "fft":
      fft = True
    finished = 0
    for u in g[v]:
      finished += dfs_util(u, dac, fft, end)
    return finished

  return dfs_util("svr", False, False, "out")

def part1(input):
  return bfs(input)

def part2(input):
  return dfs(input)

input = open("input/day11.txt", "r").read()
# input = example_input.strip()
print("PART 1 =", part1(parse_input(input)))
print("PART 2 =", part2(parse_input(input)))

File: src/rust/day11.rs
use cached::proc_macro::cached;
use std::{collections::HashMap, convert::Into};

type Graph = HashMap<String, Vec<String>>;

fn parse_input(input: &str) -> Graph {
    input
        .lines()
        .map(|line| {
            let (node, neighbors) = line.split_once(':').unwrap();
            let neighbors = neighbors.split_whitespace().map(Into::into).collect();
            (node.to_string(), neighbors)
        })
        .collect()
}

#[cached(
    key = "(String, bool, bool, String)",
    convert = r#"{ (node.to_string(), dac, fft, end.to_string()) }"#
)]
fn dfs(graph: &Graph, node: &str, mut dac: bool, mut fft: bool, end: &str) -> usize {
    if node == end {
        return (dac && fft).into();
    }
    if node == "dac" {
        dac = true;
    }
    if node == "fft" {
        fft = true;
    }
    graph
        .get(node)
        .unwrap()
        .iter()
        .map(|neighbor| dfs(graph, neighbor, dac, fft, end))
        .sum()
}

pub fn find_all_paths(input: &str) -> usize {
    let graph = parse_input(input);
    dfs(&graph, "you", true, true, "out")
}

pub fn find_all_paths_with_dac_and_fft(input: &str) -> usize {
    let graph = parse_input(input);
    dfs(&graph, "svr", false, false, "out")
}

pub fn main() {
    let input = std::fs::read_to_string("input/day11.txt").expect("Input file not found");
    let now = std::time::Instant::now();
    println!("PART 1 = {}", find_all_paths(&input));
    println!("PART 2 = {}", find_all_paths_with_dac_and_fft(&input));
    println!("Execution time: {:?}", now.elapsed());
}
